designs/spot-consolidation.md

```diff
@@ -0,0 +1,106 @@
+# Spot Consolidation
+
+## Problem Statement
+
+Customers face increased costs due to underutilized spot nodes. A customer with 4 x 24cpu pods would launch a 24xl (96cpu) instance type. If three of the four pods complete, the node will indefinitely run at 25% reservation until the pod completes, is able to fit elsewhere in the cluster, or additional work is scheduled to the node. Karpenter can currently "consolidate delete", by moving the pods to another node if capacity is available, but does not "consolidate replace", by launching a smaller replacement node for an undertuilized node. Customers have raised this inefficiency as a critical pain point with Karpenter: https://github.com/aws/karpenter-core/issues/763. 
+
+Karpenter’s consolidation algorithm minimizes the price of EC2 instances given pod scheduling constraints. However, price minimization presents challenges for EC2 Spot, which offers a tradeoff between price and interruption rate. Karpenter uses EC2 Spot’s PriceCapacityOptimized (PCO) allocation strategy to provide a balanced approach to this tradeoff, and doesn't always select the cheapest instance type, since it may result in immediate interruption. If Karpenter were to consolidate spot capacity based purely on cost, it would repeatedly interrupt the same instance, walking down the PCO decision ladder, until only the lowest price remained. Because of these complications, Karpenter does not currently support "spot replacement" consolidation.
```

Comment:
Should we consider as an option maintaining a "consolidation state machine" that keeps track of the lifecycle of consolidated nodes. For example:

1. Node is created as a result of net new Pending pods (i.e., a vanilla scale out scenario)
2. We create a lifecycle representation of this new infra:
  - "Infra XYZ" created at time T0
 3. At some point in the future, a consolidation criteria triggers a replacement of this node with one or more replacement nodes. For each new node launched as a result of this consolidation event:
  - "Infra XYZ" replaced at time T1

Let's use the simplest possible example of a node that is continually consolidated via a single node replacement event:

```
Name: Infra XYZ
Node: <a reference to the node that represents the current node on the cluster that has this "infra" history>
Events:
- created at time T0
- replaced at time T1
- replaced at time T2
- replaced at time T3
Lifetime: <computed property of the difference between T3 and T0>
EligibleForConsolidation: <bool that represents whether or not the rate of events is happening too frequently given the lifetime of the infra; if the rate is above a threshold, the value is false, if the rate is below, the value is true>
```

Using the above example, we could expose a single configurable property that expressed something like "node replacement tolerance", i.e., how much node thrashing is allowed assuming other scheduling constraints permit graceful rebalancing of workloads across nodes.

Is something like this worth considering?

Reply:
TLDR on the discussion here: We have talked about `minNodeLifetime` as a parameter that we might want to support as well as price thresholding. Both would allow us to achieve some of the tweaking that is being suggested here. I could see us adding these things to _all_ of consolidation as a fast-follow to this base solution that unlocks the ability for us to perform consolidations with spot.